File: leetcode/maximum-depth-of-binary-tree.java
//https://leetcode.com/problems/maximum-depth-of-binary-tree/

class Solution {
    // Provided TreeNode class
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int maxDepth(TreeNode root) {
        return searchMax(root, 0);
    }
    
    private int searchMax(TreeNode curr, int depth){
        if(curr != null){
            int max = Math.max(searchMax(curr.left, depth+1), searchMax(curr.right, depth+1));
            return max;
        }
        return depth;
    }
}

File: leetcode/maximum_depth_of_binary_tree.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/description/

from typing import Optional

# Provided TreeNode class
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        return self.traverse(root, 1)

    def traverse(self, node: Optional[TreeNode], depth: int) -> int:
        if node is None:
            return depth - 1
        return max(
            self.traverse(node.left, depth + 1), self.traverse(node.right, depth + 1)
        )
